tensorflow/core/kernels/data/experimental/directed_interleave_dataset_op_test.cc

```diff
@@ -0,0 +1,280 @@
+      /*input_dataset_params_vec=*/
+      std::vector<RangeDatasetParams>{RangeDatasetParams(0, 3, 1),
+                                      RangeDatasetParams(10, 13, 1)},
+      /*output_dtypes=*/{DT_INT64, DT_INT64},
+      /*output_shapes=*/{PartialTensorShape({}), PartialTensorShape({})},
+      /*num_input_datasets=*/2,
+      /*node_name=*/kNodeName);
+}
+
+DirectedInterleaveDatasetParams InvalidSelectorOuputShape() {
+  auto selector_input_dataset_params = TensorSliceDatasetParams(
+      /*components=*/{CreateTensor<int64>(TensorShape{6, 1},
+                                          {0, 1, 0, 1, 0, 1})},
+      /*node_name=*/"tensor_slice");
+  return DirectedInterleaveDatasetParams(
+      selector_input_dataset_params,
+      /*input_dataset_params_vec=*/
+      std::vector<RangeDatasetParams>{RangeDatasetParams(0, 3, 1),
+                                      RangeDatasetParams(10, 13, 1)},
+      /*output_dtypes=*/{DT_INT64, DT_INT64},
+      /*output_shapes=*/{PartialTensorShape({}), PartialTensorShape({})},
+      /*num_input_datasets=*/2,
+      /*node_name=*/kNodeName);
+}
+
+DirectedInterleaveDatasetParams InvalidSelectorValues() {
+  auto selector_input_dataset_params = TensorSliceDatasetParams(
+      /*components=*/{CreateTensor<int64>(TensorShape{6}, {2, 1, 0, 1, 0, 1})},
+      /*node_name=*/"tensor_slice");
+  return DirectedInterleaveDatasetParams(
+      selector_input_dataset_params,
+      /*input_dataset_params_vec=*/
+      std::vector<RangeDatasetParams>{RangeDatasetParams(0, 3, 1),
+                                      RangeDatasetParams(10, 13, 1)},
+      /*output_dtypes=*/{DT_INT64, DT_INT64},
+      /*output_shapes=*/{PartialTensorShape({}), PartialTensorShape({})},
+      /*num_input_datasets=*/2,
+      /*node_name=*/kNodeName);
+}
+
+DirectedInterleaveDatasetParams InvalidInputDatasetsDataType() {
+  auto selector_input_dataset_params = TensorSliceDatasetParams(
+      /*components=*/{CreateTensor<int64>(TensorShape{6}, {0, 1, 0, 1, 0, 1})},
+      /*node_name=*/"tensor_slice");
+  return DirectedInterleaveDatasetParams(
+      selector_input_dataset_params,
+      /*input_dataset_params_vec=*/
+      std::vector<RangeDatasetParams>{
+          RangeDatasetParams(0, 3, 1, {DT_INT32}),
+          RangeDatasetParams(10, 13, 1, {DT_INT64})},
+      /*output_dtypes=*/{DT_INT64, DT_INT64},
+      /*output_shapes=*/{PartialTensorShape({}), PartialTensorShape({})},
+      /*num_input_datasets=*/2,
+      /*node_name=*/kNodeName);
+}
+
+std::vector<GetNextTestCase<DirectedInterleaveDatasetParams>>
+GetNextTestCases() {
+  return {{/*dataset_params=*/DirectedInterleaveDatasetParams1(),
+           /*expected_outputs=*/{CreateTensors<int64>(
+               TensorShape({}), {{0}, {10}, {1}, {11}, {2}, {12}})}},
+          {/*dataset_params=*/DirectedInterleaveDatasetParams2(),
+           /*expected_outputs=*/{CreateTensors<int64>(
+               TensorShape({}), {{0}, {10}, {1}, {11}, {12}})}}};
```

Comment:
As required [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/experimental_dataset_ops.cc#L299), 0 data input datasets may not be allowed?

Reply:
If it isn't too much trouble, we could verify that trying to use 0 data input datasets produces a reasonable error message, as opposed to e.g. segfaulting.

Reply:
@aaudiber  Thanks for the suggestions!
> some other cases worth covering:
> 
> * Interleave with 0 data input datasets

As required [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/experimental_dataset_ops.cc#L299), 0 data input datasets may not be allowed?

> * Interleave with 1 data input dataset

Added it.

> * Interleave where num_input_datasets is set too low/too high.

Could you please explain more about this? Do you mean `num_input_datasets` is set to be larger and smaller than the size of `input_dataset_params_vec`?

Reply:
Exactly, basically a mismatch between those two parameters

Reply:
Got it. I can add these two cases. Note that [DatasetOpsTestBase::CheckOpKernelInput](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/data/dataset_test_base.cc#L530-L537) will check if the kernel inputs are valid. If we make those two parameters mismatch, the error message will come from the test base class instead of the kernel implementation.

Reply:
> If it isn't too much trouble, we could verify that trying to use 0 data input datasets produces a reasonable error message, as opposed to e.g. segfaulting.

`ZeroInputDatasetParams` test case is added. The error message will be as below
```
Invalid argument: Value for attr 'N' of 0 must be at least minimum 1; NodeDef: {{node directed_interleave_dataset}}; Op<name=DirectedInterleaveDataset; signature=selector_input_dataset:variant, data_input_datasets:N*variant -> handle:variant; attr=output_types:list(type),min=1; attr=output_shapes:list(shape),min=1; attr=N:int,min=1>

```